contracts/BaseWrapper.sol

```diff
@@ -0,0 +1,207 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity ^0.6.12;
+pragma experimental ABIEncoderV2;
+
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import {Math} from "@openzeppelin/contracts/math/Math.sol";
+import {SafeMath} from "@openzeppelin/contracts/math/SafeMath.sol";
+
+import {VaultAPI} from "./BaseStrategy.sol";
+
+interface RegistryAPI {
+    function governance() external view returns (address);
+
+    function latestVault(address token) external view returns (address);
+
+    function nextDeployment(address token) external view returns (uint256);
+
+    function vaults(address token, uint256 deploymentId) external view returns (address);
+}
+
+abstract contract BaseWrapper {
+    using Math for uint256;
+    using SafeMath for uint256;
+    using SafeERC20 for IERC20;
+
+    IERC20 public token;
+
+    // Reduce number of external calls (SLOADs stay the same)
+    VaultAPI[] private _cachedVaults;
+
+    RegistryAPI public registry;
+
+    constructor(address _token, address _registry) public {
+        token = IERC20(_token);
+        // v2.registry.ychad.eth
+        registry = RegistryAPI(_registry);
+    }
+
+    function setRegistry(address _registry) external {
+        require(msg.sender == registry.governance());
+        // In case you want to override the registry instead of re-deploying
+        registry = RegistryAPI(_registry);
+    }
+
+    function bestVault() public virtual view returns (VaultAPI) {
+        return VaultAPI(registry.latestVault(address(token)));
+    }
+
+    function allVaults() public virtual view returns (VaultAPI[] memory) {
+        uint256 cache_length = _cachedVaults.length;
+        uint256 num_deployments = registry.nextDeployment(address(token));
+
+        // Use cached
+        if (cache_length == num_deployments) {
+            return _cachedVaults;
+        }
+
+        VaultAPI[] memory vaults = new VaultAPI[](num_deployments);
+
+        for (uint256 deployment_id = 0; deployment_id < cache_length; deployment_id++) {
+            vaults[deployment_id] = _cachedVaults[deployment_id];
+        }
+
+        for (uint256 deployment_id = cache_length; deployment_id < num_deployments; deployment_id++) {
+            vaults[deployment_id] = VaultAPI(registry.vaults(address(token), deployment_id));
+        }
+
+        return vaults;
+    }
+
+    function _updateVaultCache(VaultAPI[] memory vaults) internal {
+        if (vaults.length > _cachedVaults.length) {
+            _cachedVaults = vaults;
+        }
+    }
+
+    function totalVaultBalance(address account) public view returns (uint256 balance) {
+        VaultAPI[] memory vaults = allVaults();
+
+        for (uint256 id = 0; id < vaults.length; id++) {
+            balance = balance.add(vaults[id].balanceOf(account).mul(vaults[id].pricePerShare()).div(10**uint256(vaults[id].decimals())));
```

Comment:
```suggestion
        for (uint256 i = 0; i < vaults.length; i++) {
            balance = balance.add(vaults[i].balanceOf(account).mul(vaults[i].pricePerShare()).div(10**uint256(vaults[i].decimals())));

```
why `id`? should be either `index` or `i` right?

Reply:
"deployment ID" is the index used here